File: frontend/lib/api/tasks.ts
/* ─────────────────────────────────────────────────────────────────────────────
   Unified Task Management API Client
   ───────────────────────────────────────────────────────────────────────────── */

import { api } from "@/lib/api";
import type {
  TaskCreate,
  TaskOut,
  TaskUpdate,
  TaskReportCreate,
  TaskReportOut,
  TaskBoardResponse,
} from "@/types/tasks";

export async function fetchTasks(params?: {
  task_type?: string;
  status?: string;
  patient_id?: number;
  assignee_user_id?: number;
  date_from?: string;
  date_to?: string;
  shift_date?: string;
  is_active?: boolean;
  limit?: number;
}): Promise<TaskOut[]> {
  const query = new URLSearchParams();
  if (params?.task_type) query.set("task_type", params.task_type);
  if (params?.status) query.set("status", params.status);
  if (typeof params?.patient_id === "number") query.set("patient_id", String(params.patient_id));
  if (typeof params?.assignee_user_id === "number") query.set("assignee_user_id", String(params.assignee_user_id));
  if (params?.date_from) query.set("date_from", params.date_from);
  if (params?.date_to) query.set("date_to", params.date_to);
  if (params?.shift_date) query.set("shift_date", params.shift_date);
  if (typeof params?.is_active === "boolean") query.set("is_active", String(params.is_active));
  if (typeof params?.limit === "number") query.set("limit", String(params.limit));
  
  const suffix = query.toString();
  // Trailing slash avoids FastAPI/Starlette 307 redirect from /api/tasks -> /api/tasks/
  return api.get<TaskOut[]>(suffix ? `/tasks/?${suffix}` : "/tasks/");
}

export async function fetchTaskBoard(shiftDate?: string): Promise<TaskBoardResponse> {
  const query = shiftDate ? `?shift_date=${encodeURIComponent(shiftDate)}` : "";
  return api.get<TaskBoardResponse>(`/tasks/board${query}`);
}

export async function fetchTask(taskId: number): Promise<TaskOut> {
  return api.get<TaskOut>(`/tasks/${taskId}`);
}

export async function createTask(data: TaskCreate): Promise<TaskOut> {
  return api.post<TaskOut>("/tasks/", data);
}

export async function updateTask(taskId: number, data: TaskUpdate): Promise<TaskOut> {
  return api.patch<TaskOut>(`/tasks/${taskId}`, data);
}

export async function deleteTask(taskId: number): Promise<void> {
  return api.delete<void>(`/tasks/${taskId}`);
}

export async function submitTaskReport(taskId: number, data: TaskReportCreate): Promise<TaskReportOut> {
  return api.post<TaskReportOut>(`/tasks/${taskId}/reports`, data);
}

export async function fetchTaskReports(taskId: number): Promise<TaskReportOut[]> {
  return api.get<TaskReportOut[]>(`/tasks/${taskId}/reports`);
}

/** Browser URL for streaming a pending attachment (create-task flow; cookie auth). */
export function taskPendingAttachmentContentUrl(pendingId: string): string {
  return `/api/tasks/attachments/pending/${encodeURIComponent(pendingId)}/content`;
}

/** Browser URL for a finalized task template / subtask attachment. */
export function taskTemplateAttachmentContentUrl(
  taskId: number,
  attachmentId: string,
): string {
  return `/api/tasks/${taskId}/attachments/${encodeURIComponent(attachmentId)}/content`;
}

export async function resetRoutineTasks(shiftDate?: string): Promise<{ reset_count: number }> {
  const query = shiftDate ? `?shift_date=${encodeURIComponent(shiftDate)}` : "";
  return api.post<{ reset_count: number }>(`/tasks/routines/reset${query}`, {});
}

File: server/app/api/endpoints/tasks.py
from __future__ import annotations

from datetime import date, datetime, timezone
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import FileResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.dependencies import (
    get_current_active_user,
    get_current_user_workspace,
    get_db,
    get_visible_patient_ids,
)
from app.models.core import Workspace
from app.models.users import User
from app.schemas.tasks import (
    TaskBoardResponse,
    TaskCreate,
    TaskOut,
    TaskReportCreate,
    TaskReportOut,
    TaskUpdate,
)
from app.services.tasks import task_service
from app.services.workflow_message_attachments import (
    read_pending_attachment_bytes,
    resolve_attachment_from_task_json,
)

router = APIRouter()

_HEAD_NURSE_ADMIN = frozenset({"admin", "head_nurse"})
_TASK_EXECUTORS = frozenset({"admin", "head_nurse", "supervisor", "observer"})


def _require_management_role(user: User) -> None:
    """Raise 403 if user is not head_nurse or admin."""
    if user.role not in _HEAD_NURSE_ADMIN:
        raise HTTPException(403, detail="Only head nurse or admin can manage tasks")


def _require_executor_role(user: User) -> None:
    """Raise 403 if user is not a valid task executor."""
    if user.role not in _TASK_EXECUTORS:
        raise HTTPException(403, detail="You do not have permission to execute tasks")


@router.get("/", response_model=list[TaskOut])
async def list_tasks(
    task_type: Optional[str] = Query(None, description="Filter by task type: specific or routine"),
    status: Optional[str] = Query(None, description="Filter by status"),
    patient_id: Optional[int] = Query(None, description="Filter by patient ID"),
    assignee_user_id: Optional[int] = Query(None, description="Filter by assignee user ID"),
    date_from: Optional[datetime] = Query(None, description="Filter by created_at >= date_from"),
    date_to: Optional[datetime] = Query(None, description="Filter by created_at <= date_to"),
    shift_date: Optional[date] = Query(None, description="Filter by shift date"),
    is_active: bool = Query(True, description="Filter by active state"),
    limit: int = Query(100, ge=1, le=500),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_active_user),
    workspace: Workspace = Depends(get_current_user_workspace),
):
    """List tasks with filtering. Enforces workspace and patient visibility."""
    ws_id = workspace.id
    
    visible_patient_ids = await get_visible_patient_ids(db, ws_id, user)
    
    return await task_service.list_tasks(
        db,
        ws_id,
        user_id=user.id,
        user_role=user.role,
        visible_patient_ids=visible_patient_ids,
        task_type=task_type,
        status=status,
        patient_id=patient_id,
        assignee_user_id=assignee_user_id,
        date_from=date_from,
        date_to=date_to,
        shift_date=shift_date,
        is_active=is_active,
        limit=limit,
    )


@router.get("/board", response_model=TaskBoardResponse)
async def get_task_board(
    shift_date: Optional[date] = Query(None, description="Shift date (defaults to today)"),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_active_user),
    workspace: Workspace = Depends(get_current_user_workspace),
):
    """Get task board with per-user aggregation."""
    ws_id = workspace.id
    visible_patient_ids = await get_visible_patient_ids(db, ws_id, user)
    
    return await task_service.get_task_board(
        db,
        ws_id,
        user_id=user.id,
        user_role=user.role,
        visible_patient_ids=visible_patient_ids,
        shift_date=shift_date,
    )


@router.get("/attachments/pending/{pending_id}/content")
async def stream_pending_task_attachment(
    pending_id: str,
    user: User = Depends(get_current_active_user),
    workspace: Workspace = Depends(get_current_user_workspace),
):
    """Stream a pending upload for preview (same storage as workflow message pending uploads)."""
    path, media_type, filename = read_pending_attachment_bytes(
        workspace_id=workspace.id,
        user_id=user.id,
        pending_id=pending_id,
    )
    return FileResponse(
        path,
        media_type=media_type,
        headers={"Content-Disposition": f'inline; filename="{filename}"'},
    )


@router.get("/{task_id}/attachments/{attachment_id}/content")
async def stream_task_template_attachment(
    task_id: int,
    attachment_id: str,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_active_user),
    workspace: Workspace = Depends(get_current_user_workspace),
):
    """Download/preview an attachment embedded in task report template or subtask report_spec."""
    ws_id = workspace.id
    visible_patient_ids = await get_visible_patient_ids(db, ws_id, user)
    task = await task_service.get_task(
        db,
        ws_id,
        task_id,
        user_id=user.id,
        user_role=user.role,
        visible_patient_ids=visible_patient_ids,
    )
    if not task:
        raise HTTPException(404, detail="Task not found")
    path, media_type, filename = resolve_attachment_from_task_json(
        task.report_template,
        task.subtasks,
        attachment_id,
    )
    return FileResponse(
        path,
        media_type=media_type,
        headers={"Content-Disposition": f'inline; filename="{filename}"'},
    )


@router.get("/{task_id}", response_model=TaskOut)
async def get_task(
    task_id: int,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_active_user),
    workspace: Workspace = Depends(get_current_user_workspace),
):
    """Get single task by ID."""
    ws_id = workspace.id
    visible_patient_ids = await get_visible_patient_ids(db, ws_id, user)
    
    task = await task_service.get_task(
        db,
        ws_id,
        task_id,
        user_id=user.id,
        user_role=user.role,
        visible_patient_ids=visible_patient_ids,
    )
    
    if not task:
        raise HTTPException(404, detail="Task not found")
    
    return task


@router.post("/", response_model=TaskOut, status_code=201)
async def create_task(
    obj_in: TaskCreate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_active_user),
    workspace: Workspace = Depends(get_current_user_workspace),
):
    """Create new task. Requires head_nurse or admin role."""
    _require_management_role(user)
    ws_id = workspace.id
    
    return await task_service.create_task(
        db,
        ws_id,
        actor_user_id=user.id,
        actor_user_role=user.role,
        obj_in=obj_in,
    )


@router.patch("/{task_id}", response_model=TaskOut)
async def update_task(
    task_id: int,
    obj_in: TaskUpdate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_active_user),
    workspace: Workspace = Depends(get_current_user_workspace),
):
    """Update task. Requires head_nurse or admin role."""
    _require_management_role(user)
    ws_id = workspace.id
    
    task = await task_service.update_task(
        db,
        ws_id,
        task_id,
        actor_user_id=user.id,
        actor_user_role=user.role,
        obj_in=obj_in,
    )
    
    if not task:
        raise HTTPException(404, detail="Task not found")
    
    return task


@router.delete("/{task_id}", status_code=204)
async def delete_task(
    task_id: int,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_active_user),
    workspace: Workspace = Depends(get_current_user_workspace),
):
    """Soft delete task. Requires head_nurse or admin role."""
    _require_management_role(user)
    ws_id = workspace.id
    
    deleted = await task_service.delete_task(
        db,
        ws_id,
        task_id,
        actor_user_id=user.id,
        actor_user_role=user.role,
    )
    
    if not deleted:
        raise HTTPException(404, detail="Task not found")


@router.post("/{task_id}/reports", response_model=TaskReportOut, status_code=201)
async def submit_report(
    task_id: int,
    obj_in: TaskReportCreate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_active_user),
    workspace: Workspace = Depends(get_current_user_workspace),
):
    """Submit structured task report. Must be assignee or head_nurse/admin."""
    _require_executor_role(user)
    ws_id = workspace.id
    
    return await task_service.submit_report(
        db,
        ws_id,
        task_id,
        submitter_user_id=user.id,
        submitter_user_role=user.role,
        obj_in=obj_in,
    )


@router.get("/{task_id}/reports", response_model=list[TaskReportOut])
async def get_reports(
    task_id: int,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_active_user),
    workspace: Workspace = Depends(get_current_user_workspace),
):
    """Get all reports for a task. Requires task access."""
    ws_id = workspace.id
    visible_patient_ids = await get_visible_patient_ids(db, ws_id, user)
    
    # Verify user can see the task first
    task = await task_service.get_task(
        db,
        ws_id,
        task_id,
        user_id=user.id,
        user_role=user.role,
        visible_patient_ids=visible_patient_ids,
    )
    
    if not task:
        raise HTTPException(404, detail="Task not found or access denied")
    
    return await task_service.get_task_reports(db, ws_id, task_id)


@router.post("/routines/reset")
async def reset_routine_tasks(
    shift_date: Optional[date] = Query(None, description="Shift date to reset (defaults to today)"),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_active_user),
    workspace: Workspace = Depends(get_current_user_workspace),
):
    """Reset all routine tasks for a shift date. Requires head_nurse or admin role."""
    _require_management_role(user)
    ws_id = workspace.id
    
    reset_count = await task_service.reset_routine_tasks(
        db,
        ws_id,
        actor_user_id=user.id,
        actor_user_role=user.role,
        target_shift_date=shift_date,
    )
    
    return {"reset_count": reset_count}
